README.md
# Sunset to Sunset
This script was developed to help with closing down of a website from sunset on Friday to sunset on Saturday. It displays a banner leading up to the Sabbath notifying visitors of the site when it will not be available.

## Features
- Automatically show a banner on the top of your site notifying visitors when your site will close.
- Show a message during the Sabbath hours letting visitors know why you are closed and when your site will be back online.

## Installation
Whichever method you choose, either **Download** or **CDN**, it’s best to include all these files in the head so that they load right away. It’s not ideal to have render-blocking scripts in the head but the reason we recommend putting them in the head is to avoid the flash of the rendered page before the Sabbath message gets rendered. Feel free to submit a PR for improved installation methods.

### Download
- **CSS**
  - [style.css](https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/style.css), minified
- **Javascript**
  - [polyfills-legacy.min.js](https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/polyfills-legacy.min.js), legacy bundle for outdated browser support.
  - [sunset-to-sunset.min.js](https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/sunset-to-sunset.min.js), minified
  - [sunset-to-sunset-legacy.min.js](https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/sunset-to-sunset-legacy.min.js), minified for legacy browsers

``` html
<!DOCTYPE html>
<html lang="en">

<head>
  <!-- ... -->

  <!-- Sunset to Sunset javascript -->
  <script type="module" src="YOUR/PATH/HERE/sunset-to-sunset.min.js" crossorigin></script>

  <!-- Sunset to Sunset stylesheet -->
  <link href="YOUR/PATH/HERE/style.css" rel="stylesheet" media="print" onload="this.media='all'">

  <!-- Legacy bundle for outdated browser support. -->
  <script type="nomodule" src="YOUR/PATH/HERE/polyfills-legacy.min.js"></script>
  
  <!-- Sunset to Sunset javascript for legacy browsers -->
  <script type="nomodule" src="YOUR/PATH/HERE/sunset-to-sunset-legacy.min.js"></script>

</head>

<body>
```

### CDN
Link directly to the Sunset to Sunset files on [unpkg](https://unpkg.com/).
``` html
<!DOCTYPE html>
<html lang="en">

<head>
  <!-- ... -->

  <!-- Sunset to Sunset javascript -->
  <script
    type="module"
    src="https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/sunset-to-sunset.min.js"
    crossOrigin="true"
  ></script>

  <!-- Sunset to Sunset stylesheet -->
  <link
    href="https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/style.css"
    rel="stylesheet"
  />

  <!-- Legacy bundle for outdated browser support. -->
  <script
    noModule
    crossOrigin="true"
    id="vite-legacy-polyfill"
    src="https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/polyfills-legacy.min.js"
  ></script>

  <!-- Sunset to Sunset javascript for legacy browsers -->
  <script
    noModule
    crossOrigin="true"
    id="vite-legacy-entry"
    data-src="https://unpkg.com/sunset-to-sunset@1.0.7/dist/assets/sunset-to-sunset-legacy.min.js"
  >
    System.import(
    document.getElementById('vite-legacy-entry').getAttribute('data-src'),
    );
  </script>

</head>

<body>
```

### Support for Safari 10.1
Safari 10.1 supports modules, but does not support the `nomodule` attribute - it will load `<script nomodule>` anyway. If you need to support Safari 10.1 on your website you may use this snippet below:

```html
<!-- Safari 10.1 `nomodule` fix script (https://gist.github.com/samthor/64b114e4a4f539915a95b91ffd340acc) -->
<!-- This can be omitted if you don't have many users of Safari 10.1 in your target audience. -->
<script>
  !function(){var e=document,t=e.createElement("script");if(!("noModule"in t)&&"onbeforeload"in t){var n=!1;e.addEventListener("beforeload",function(e){if(e.target===t)n=!0;else if(!e.target.hasAttribute("nomodule")||!n)return;e.preventDefault()},!0),t.type="module",t.src=".",e.head.appendChild(t),t.remove()}}();
</script>
```


## Usage
Sunset to sunset needs some initial configuration needed for it to work correctly for your location. It will work out of the box without configuration but it won't be for your location.

## Custom Templates
Sunset to sunset has some built-in banner and message templates that will be shown at the appropriate times but sometimes you need to define your own wording. You can do that with `<div>`s with the appropriate `id`s added to to them.

### Special Template Tags
You can add the following two tags to your custom templates to render the closing time and the opening times.

#### Closing Time
``` html
<span class="sts-closing-time"></span>
```
#### Opening Time
``` html
<span class="sts-opening-time"></span>
```

### Formatting Times
There are three methods for setting the format on the calculated times in your custom templates:

1. **[Token](#token):** Allows you to set what order each part of the date and time appear so it looks the same for everyone.
2. **[Locale](#locale):** Using whatever the user's browser's date/time format is set to.
3. **[Default](#default):** Basically let the script decide for you.

#### Token
This is the most flexible formatting option as it allows you to define the order of the date/time parts and the formatting. To use the `token` format include the [special closing and opening time tags](#special-template-tags) with a `data-format-token` like the following:
``` html
<span class="sts-closing-time" data-format-token="cccc 'at' h:mm a ZZZZ"></span>

<!-- Output: Monday at 7:44 PM CDT -->
```
> Note that you can add words in the token format by surrounding them with single quotes like in the example above.

**Advanced formatting:** The [Luxon documentation has a full list of formatting tokens](https://moment.github.io/luxon/#/formatting?id=table-of-tokens) so that you can fine-tune your dates and times.

#### Locale
With this option you can't arrange the parts of the date and time but you can decide how they are formatted. To use the `locale` format include the [special closing and opening time tags](#special-template-tags) with a `data-format-locale` like the following:
``` html
<span class="sts-closing-time" data-format-locale='{ 
  "weekday": "long",
  "month": "long",
  "hour": "numeric",
  "minute": "numeric",
  "timeZoneName": "short"
}'></span>

<!-- Output: September Monday, 7:44 PM CDT -->
```

#### Default
Include the [special closing and opening time tags](#special-template-tags) with no `data-format-locale` or `data-format-token` attributes and it will output the times with the token format of `cccc, LLLL d 'at' h:mm a ZZZZ`:
``` html
<span class="sts-closing-time"></span>

<!-- Output: Monday, September 27 at 7:44 PM CDT -->
```

### Banner Template
To define your custom banner template add the following snippet to your page, preferrably in the `<head>` tag.


``` html
<div id="sts-banner-template">
  <div class="sts-banner  YOUR-CUSTOM-CLASSES-HERE">
    Our store will be closing at 
    <span class="sts-closing-time"></span>
    <span>and will re-open on</span>
    <span class="sts-opening-time"></span>
  </div>
</div>
```
The `sts-banner` class provides some default styling for the banner that you could use as a starting point. Or you can remove it altogether and use your own custom styles. If you keep the `sts-banner` class it will output something like what you see below:

![banner screenshot](src/screenshots/banner.png)

### Simple Message Template
If you want to just set a paragraph or two of text to appear during Sabbath then this is the template you'll want to use. 

To define your custom message template add the following snippet to your page, preferrably in the `<head>` tag:
``` html
<div id="sts-message-template">
  <p>
    Your message here
  </p>
</div>
```

It will output something like what you see below:

![simple custom message template screenshot](src/screenshots/simple-custom-message.png)

If you supply no custom template then a default message will be set which looks like this:

![simple message template screenshot](src/screenshots/simple-default-message.png)


### Full Message Template
If you want to have full control over the appearance and layout of the message then you'll want to use the full message template. You will need to supply your own custom styling to add padding and centering of your message.

To define your custom message template add the following snippet to your page, preferrably in the `<head>` tag:
``` html
<div id="sts-full-message-template">
  <div class="your-custom-layout-class">
    <div class="your-message-area">
      <h1 class="your-message__heading">Sabbath</h1>
      <p>Message here</p>
    </div>
    <div class="your-custom-time-area">
      <p>
        We will re-open on <strong><span class="sts-opening-time"></span>.</strong>
      </p>
    </div>
  </div>
</div>
```

## Defining the Settings
The settings are defined with an html `template` element. The template must have an `id` of `sts-settings` and the data attribute `data-settings` like below:

``` html
<template 
  id="sts-settings" 
  data-settings='{
    "location": {
      "lat": 35.7686615,
      "long": -87.4871698
    }
}'></template>
```

> **N.B.:** the `data-settings` attribute must be valid JSON. Keys need to be quoted, for example `"location":`. The attribute value uses single quotes ', but the JSON entities use double-quotes ".

## Settings
You can pass an object of configuration options with the `data-settings` attribute in the template above. The allowed values are as follows:

### `location`
- **Type:** `Object`
- **Default:** `{ "lat": 0, "long": 0 }`
- **Description:** An object defining the latitude and longitude of the location to calculate the sunset times. Both keys `lat` and `long` are expected to be of the `Number` type.
- **Example:**
  ``` json
  "location": {
    "lat": 35.7686615,
    "long": -87.4871698
  }
  ```

### `guardDuration`
- **Type:** `Object`
- **Default:** `{ "minutes": 30 }`
- **Description:** This allows you to set the duration of the guard before and after the Sabbath. Whatever time you set here will determine when your Sabbath message will come up and go down. 
  - It accepts any object of options that can have any of the following keys: `years`, `quarters` (three months), `months`, `weeks`, `days`, `hours`, `minutes`, `seconds`, and `milliseconds`. It's recommended to the smaller units like `hours`, `minutes`, etc. though because otherwise you'll be calculating guard times that are into the next week.
- **Example:** If the sun set on Friday at 8:00pm and on Saturday at 8:02pm this would calculate the closing guard time at 5:15pm on Friday and the opening time at 10:47pm on Saturday:
  ``` json
  "guardDuration": {
    "hours": 2,
    "minutes": 45
  }
  ```

### `bannerDuration`
- **Type:** `Object`
- **Default:** `{ "hours": 3 }`
- **Description:** This allows you to set the duration that the banner will be visible before the Sabbath message appears.
  - It accepts any object of options that can have any of the following keys: `years`, `quarters` (three months), `months`, `weeks`, `days`, `hours`, `minutes`, `seconds`, and `milliseconds`. Generally you will only need to use `hours` and `minutes` though.
- **Example:** This would show the banner 3 hours and 3 minutes before the calculated closing guard time determined by the `guardDuration` option:
  ``` json
  "bannerDuration": {
    "hours": 3,
    "minutes": 30
  }
  ```

### `simulateTime`
- **Type:** `Boolean | String`
- **Default:** `false`
- **Description:** This allows you to simulate the time to see how the plugin works at different times of the week and different times of the day. The accepted values are:
  - `preparation-day`
  - `banner-up`
  - `during-sabbath`
  - `after-sabbath`
- **Example:** This would show the Sabbath message as it would appear during the Sabbath hours.
  ``` json
  "simulateTime": "during-sabbath"
  ```

### `debug`
- **Type:** `Boolean`
- **Default:** `false`
- **Description:** When this is set to `true` it will output the calculated times to the console. **N.B.:** this will only show the calculated times if the current weekday is equal to the Friday or Saturday or if the `simulateTime` setting is not `false`. This is done so that it doesn't need to calculate the sunset times unnessarily during the week.
- **Example:** This would output the calculated times to the console.
  ``` json
  "debug": "true"
  ```

CHANGELOG.md
# Sunset To Sunset Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/) and this project adheres to [Semantic Versioning](http://semver.org/).

## Unreleased
### Changed
- User-defined templates for the message that appears on Sabbath and the banner message no longer use the HTML `<template>` tag. They need to use any other element now, preferably a `<div>` but any element with the right `id` will work. This change was needed because of a conflict with using `<template>` in React apps as React expects it to be a JSX template and not the HTML `<template>` tag.

## 1.0.6 - 2022-09-12
### Changed
- Luxon is now included in the build.
- Updated dependencies.

## 1.0.5 - 2021-10-15
### Changed
- Updated the test file settings to show a more syntax-safe JSON object.

### Fixed
- Script will not hide the `html` element while it's determining if it should render a message.

## 1.0.4 - 2021-09-27
### Added
- Added `README.md`.
- Added the option to set the closing and opening days via html for debugging purposes.

### Fixed
- Fixed an issue where you really couldn't see what time it would close on the upcoming closing day when simulating the time.

## 1.0.3 - 2021-09-26
### Fixed
- Fixed an issue where the `html` tag was not being hidden once the script had loaded.

## 1.0.2 - 2021-09-26
### Fixed
- Added new build files that should have been in `1.0.1`.

## 1.0.1 - 2021-09-26
### Changed
- Specify Vite port number so that it doesn't conflict with other test projects running on port `3000`.

### Fixed
- Fix an issue where the opening times were being set too soon on the full message template and subsequently not being rendered. ([#3](https://github.com/cavellblood/sunset-to-sunset/issues/3))

## 1.0.0 - 2021-09-23
### Changed
- Initial Release

## 1.0.0-beta.2 - 2021-09-23
### Fixed
- Added build files.

## 1.0.0-beta.1 - 2021-09-23
### Added
- Added a default banner template if no template is defined.
- Allow multiple closing and opening elements to be added to the banner.
- Allow the user to set a custom simple message or a fully custom message template.

vite.config.js
import legacy from "@vitejs/plugin-legacy";
import { nodeResolve } from "@rollup/plugin-node-resolve";
import path from "path";

// https://vitejs.dev/config/
export default ({ command }) => ({
  build: {
    emptyOutDir: true,
    manifest: true,
    rollupOptions: {
      input: {
        "sunset-to-sunset": "./src/tests/index.html",
        style: "./src/css/style.css",
      },
      output: {
        sourcemap: true,
        entryFileNames: `assets/[name].min.js`,
        chunkFileNames: `assets/[name].js`,
        assetFileNames: `assets/[name].[ext]`,
      },
    },
  },
  publicDir: "./src/tests",
  plugins: [
    legacy({
      targets: ["defaults", "not IE 11"],
    }),
    nodeResolve({
      moduleDirectories: ["node_modules"],
    }),
  ],
  server: {
    port: 3100,
  },
});


src/js/index.js
const extend = (a, b) => {
	return Object.assign(a, b)
}

const SunsetToSunset = (options) => {

	console.log(`Intializing Sunset to Sunset...`);
	
	// Set default options
	const defaults = {
		storeIsOpen: true,
		showSunsetMessage: false,
		guardDuration: {
			minutes: 30
		},
		messageDuration: {
			minutes: 60
		},
		location: {
			lat: 0,
			long: 0
		}
	}
	
	// Merge options with defaults
	options = extend(extend({}, defaults), options)
	
	const DateTime = luxon.DateTime;
	const Duration = luxon.Duration;

	const now = DateTime.now()
	const today = now.toString()
	
	// Set day of week: zero-based index
	const closingDayNumber = 4
	const openingDayNumber = 6
	
	// Get location coordinates
	const getLocation = () => {
		return options.location;
	}
	
	// Get sunset time for given date and location
	const getSunsetTime = async (date) => {
		const sunsetDate = date.split('T')[0]
		const response = await fetch('https://api.sunrise-sunset.org/json?&lat=' + getLocation().lat + '&lng=' + getLocation().long + '&date=' + sunsetDate + '&formatted=0')
		const data = await response.json();

		return data.results.sunset
	}
	
	// Get closing time
	const getClosingTime = () => {
		const daysToClosing = closingDayNumber - now.weekday
		const closingDate = DateTime.fromISO(now.plus({
			days: daysToClosing
		})).toString()

		const closingTime = getSunsetTime(closingDate);

		return closingTime
	}

	// Get opening time
	const getOpeningTime = () => {
		const daysToOpening = openingDayNumber - now.weekday
		const openingDate = DateTime.fromISO(now.plus({
			days: daysToOpening
		})).toString()

		const openingTime = getSunsetTime(openingDate);

		return openingTime
	}
	
	// Get message minutes
	const getMessageDuration = () => {
		let duration = Duration.fromObject(options.messageDuration)
		
		return duration
	}
	
	// Get guart minutes
	const getGuardMinutes = () => {
		return options.guardMinutes
	}
	
	//Get message time
	const getMessageTime = (date) => {
		return DateTime.fromISO(date).minus(getMessageDuration())
	}
	
	const getGuardTime = (date, action) => {
		let time
		
		if (action == 'closing') {
			time = DateTime.fromISO(date).minus({ minutes: getGuardMinutes() })
		}
		
		if (action == 'opening') {
			time = DateTime.fromISO(date).plus({ minutes: getGuardMinutes() })
		}
		
		return time
	}
	
	const getTimes = async () => {
		const allTimes = Promise.all([
			getClosingTime(),
			getOpeningTime()
		])
		
		const times = await allTimes
		
		return times
	}
	
	// Only run if today is closing day
	if ( now.weekday == closingDayNumber ) {
		getTimes().then(([closingTime, openingTime]) => {
			
			// Set guard times
			const closing = getGuardTime(closingTime, 'closing')
			const opening = getGuardTime(openingTime, 'opening')
			
			// Check times

			// Is is during the week before closing time?
			const duringWeek    = now < closing

			// Is it before closing time but the banner should be up?
			const showBanner = now < closing && now > getMessageTime
			
			// Is it during the sabbath?
			const duringSabbath = now >= closing && now <= opening && now.weekday >= closingDayNumber
			
			// Is it after sundown on Saturday?
			const afterSabbath = now > opening && now >= openingDayNumber

			console.log(now.toString())
			console.log(getMessageTime(closing).toString())
			console.log(duringWeek)
			console.log(showBanner)
			console.log(duringSabbath)
			console.log(afterSabbath)
		})
	}
	
}


SunsetToSunset({
	location: {
		lat: 35.750413,
		long: -87.229406
	},
	messageMinutes: 5 * 60
})



src/js/main.js
import '../css/style.css'

const extend = (a, b) => {
	return Object.assign(a, b)
}

const renderBanner = (closing, opening) => {
	const userDefinedTemplate = document.querySelector('template#sts-banner-template')

	let template
	
	if (userDefinedTemplate) {
		template = userDefinedTemplate
	} else {
		let defaultBannerTemplate = document.createElement("template")
		
		defaultBannerTemplate.innerHTML = `
			<div class="sts-banner">
				Because of religious beliefs our store will be closed over the Sabbath hours.
				Closing on <span class="sts-closing-time"></span> and will re-open on <span class="sts-opening-time"></span>.
			</div>
		`
		template = defaultBannerTemplate
	}

	const banner = template.content.cloneNode(true)

	// Find all closing elements and add formatted times
	const closingElements = banner.querySelectorAll('.sts-closing-time')
	formatTimes(closingElements, closing)

	// Find all opening elements and add formatted times
	const openingElements = banner.querySelectorAll('.sts-opening-time')
	formatTimes(openingElements, opening)
	
	// Insert the banner as the first item on the page.
	document.body.insertBefore(banner, document.body.firstChild)
}

const renderMessage = (opening) => {
	const userFullTemplate = document.querySelector('template#sts-full-message-template')
	const userSimpleTemplate = document.querySelector('template#sts-message-template')
	const userDefinedTemplate = userFullTemplate ? userFullTemplate : userSimpleTemplate

	let message

	if (userDefinedTemplate) {
		message = userDefinedTemplate.content.cloneNode(true)
		message = message.firstElementChild
	} else {
		let messageHolder = document.createElement("div")
		messageHolder.innerHTML = `
			<p>
				In a world that seems to be spinning faster every day, we choose to stop 
				and rest every Sabbath (Saturday). It’s a day for us to relax, refresh, 
				refocus and worship; worship a God who loved us so much that He built a 
				day of rest into Creation week and then commanded us to keep it 
				(knowing we probably wouldn’t do it otherwise—even though it is for 
				our best good).
			</p>
		`

		message = messageHolder
	}

	let template = document.createElement("template")
	template.innerHTML = `
		<div class="sts-full-message__container">

			<div class="sts-layout  sts-modal">
				<div class="sts-layout__item  sts-message-area  relative">
					<h1 class="sts-full-message__heading">Sabbath</h1>
				</div>
				<div class="sts-layout__item  sts-time-area">
					<p>
						We will re-open on <strong><span class="sts-opening-time"></span>.</strong>
					</p>
				</div>
			</div>
		
		</div>
	`

	const messageTemplate = template.content.cloneNode(true)

	if (userFullTemplate) {
		let messageContainer = messageTemplate.querySelector('.sts-full-message__container');

		while (messageContainer.firstElementChild) {
			messageContainer.removeChild(messageContainer.firstChild)
		}

		messageContainer.insertBefore(message, null)

	} else if (userSimpleTemplate) {
		const messageArea = messageTemplate.querySelector('.sts-message-area');
		messageArea.insertBefore(message, null)
	}


	const html = document.getElementsByTagName('html')[0]
	html.classList.add('sts-during-sabbath')

	// Find all opening elements and add formatted times
	const openingElements = messageTemplate.querySelectorAll('.sts-opening-time')
	formatTimes(openingElements, opening)
	
	// Insert the messageTemplate as the first item on the page.
	document.body.insertBefore(messageTemplate, document.body.firstChild)
}

const formatTimes = (elements, time) => {
	const defaultFormats = {
		locale: {
			"weekday": "long",
			"month": "long",
			"hour": "numeric",
			"minute": "numeric",
			"timeZoneName": "short"
		},
		token: "cccc, LLLL d 'at' h:mm a ZZZZ"
	}

	for (let index = 0; index < elements.length; index++) {
		const el = elements[index];
		
		const elFormatToken  = el.dataset.formatToken
		const elFormatLocale = el.dataset.formatLocale
		let formatted
	
		if (elFormatToken) {
			// Use token format first (https://moment.github.io/luxon/#/formatting?id=table-of-tokens)
			formatted = time.toFormat(elFormatToken)
		} else if (elFormatLocale) {
			// Use locale format
			formatted = time.toLocaleString(JSON.parse(elFormatLocale))
		} else {
			formatted = time.toFormat(defaultFormats.token)
		}

		// Set the content of the element to the formatted time.
		el.textContent = formatted
	}
}

const SunsetToSunset = (() => {

	console.log(`Intializing Sunset to Sunset...`);

	// Set day of week: zero-based index
	const closingDayNumber = 5
	const openingDayNumber = 6

	const DateTime = luxon.DateTime
	const Duration = luxon.Duration

	const now = DateTime.now()

	// Keep track of calculated times
	let times = {}

	times["Current Time"] = {
		'Date': now.toLocaleString(DateTime.DATE_FULL),
		'Time': now.toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET),
	}

	const activateSunsetWatch = now.weekday == closingDayNumber || now.weekday == openingDayNumber

	const html = document.getElementsByTagName('html')[0]

	// Set default options
	const defaults = {
		guardDuration: {
			minutes: 30
		},
		messageDuration: {
			minutes: 60
		},
		location: {
			lat: 0,
			long: 0
		},
		simulateTime: false,
		debug: false
	}
	
	// Get options set in HTML
	const stsContainer = document.querySelector('template#sts-settings')
	let options = stsContainer != null ? JSON.parse(stsContainer.dataset.settings) : {}
	
	// Merge options with defaults
	options = extend(extend({}, defaults), options)

	// Hide `html` until we have determined what things need to be rendered.
	if (activateSunsetWatch || options.simulateTime) {
		html.style.display = "none";
	}

	if (options.simulateTime) {
		console.warn('%cThe `simulateTime` option is enabled for the Sunset to Sunset plugin. Remember to disable this option once you are done verifying the settings.', 'font-size: 16px')
	}
	
	if (options.debug) {
		console.group(`Sunset to Sunset intialized with the following options:`)
		console.dir(options)
		console.groupEnd()
	}
	
	// Get location coordinates
	const getLocation = () => {
		return options.location;
	}
	
	// Get sunset time for given date and location
	const getSunsetTime = async (date) => {
		const sunsetDate = date.split('T')[0]
		const response = await fetch('https://api.sunrise-sunset.org/json?&lat=' + getLocation().lat + '&lng=' + getLocation().long + '&date=' + sunsetDate + '&formatted=0')
		const data = await response.json();
		const sunset = DateTime.fromISO(data.results.sunset)

		return sunset
	}
	
	// Get closing sunset
	const getClosingSunset = () => {
		let daysToClosing

		
		if (options.simulateTime) {
			daysToClosing = 0
		} else {
			daysToClosing = closingDayNumber - now.weekday
		}

		const closingDate = DateTime.fromISO(now.plus({
			days: daysToClosing
		})).toString()

		const closingSunset = getSunsetTime(closingDate);

		return closingSunset
	}

	// Get opening sunset
	const getOpeningSunset = () => {
		let daysToOpening

		if (options.simulateTime) {
			daysToOpening = 1
		} else {
			daysToOpening = openingDayNumber - now.weekday
		}

		const openingDate = DateTime.fromISO(now.plus({
			days: daysToOpening
		})).toString()

		const openingSunset = getSunsetTime(openingDate);

		return openingSunset
	}
	
	// Get message minutes
	const getMessageDuration = () => {
		let duration = Duration.fromObject(options.messageDuration)
		
		return duration
	}
	
	// Get guard duration
	const getGuardDuration = () => {
		return options.guardDuration
	}
	
	//Get message time
	const getMessageTime = (date) => {
		return date.minus(getMessageDuration())
	}
	
	// Get guard time. `date` is a Luxon DateTime object.
	const getGuardTime = (date, action) => {
		let time
		
		if (action == 'closing') {
			time = date.minus(getGuardDuration())
		}
		
		if (action == 'opening') {
			time = date.plus(getGuardDuration())
		}
		
		return time
	}
	
	const getTimes = async () => {
		const allTimes = Promise.all([
			getClosingSunset(),
			getOpeningSunset()
		])
		
		const times = await allTimes
		
		return times
	}
	
	// Only run if today is Friday or Sabbath
	if ( activateSunsetWatch || options.simulateTime) {
		getTimes().then(([closingSunset, openingSunset]) => {
			
			// Set guard times
			const closing = getGuardTime(closingSunset, 'closing')
			const opening = getGuardTime(openingSunset, 'opening')

			let preparationDay = false
			let bannerUp = false
			let duringSabbath = false
			let afterSabbath = false

			// Check if we are simulating the time
			if (options.simulateTime) {
				switch (options.simulateTime) {
					case 'preparation-day':
						preparationDay = true
						break;

					case 'banner-up':
						bannerUp = true
						break;

					case 'during-sabbath':
						duringSabbath = true
						break;

					case 'after-sabbath':
						afterSabbath = true
						break;
						
					default:
						break;
				}
			} else {
				preparationDay = now < getMessageTime(closing)
				bannerUp = now > getMessageTime(closing) && now < closing
				duringSabbath = now >= closing && now <= opening && now.weekday >= closingDayNumber
				afterSabbath = now > opening && now >= openingDayNumber
			}
		
			// Is is during the week before the time to show the banner?
			if (preparationDay) {
				// Refresh the page when it's time to show the banner.
				const refreshTime = getMessageTime(closing).diff(now, 'milliseconds').toObject();

				// Don't refresh if there is a negative refresh time.
				// This conditional is mainly here for when `simulateTime`
				// is set to `preparation-day`.
				if (refreshTime.milliseconds >= 0) {
					setTimeout(() => {
						location.reload()
					}, refreshTime.milliseconds);
				}
			}

			// Is it before closing time but the banner should be up?
			if (bannerUp) {
				renderBanner(closing, opening)

				// Refresh the page when it's closing time.
				const refreshTime = closing.diff(now, 'milliseconds').toObject();
				
				// Don't refresh if there is a negative refresh time.
				// This conditional is mainly here for when `simulateTime`
				// is set to `banner-up`.
				if (refreshTime.milliseconds >= 0) {
					setTimeout(() => {
						location.reload()
					}, refreshTime.milliseconds);
				}
			}

			// Is it during the sabbath?
			if (duringSabbath) {
				renderMessage(opening)

				// Refresh the page when it's opening time.
				const refreshTime = opening.diff(now, 'milliseconds').toObject();

				if (refreshTime.milliseconds >= 0) {
					setTimeout(() => {
						location.reload()
					}, refreshTime.milliseconds);
				}
			}

			// Is it after sundown on Saturday?
			if (afterSabbath) {

			}

			// Unhide `html` after we determine what things need to be rendered.
			html.style.display = "block"

			if (options.debug) {

				let checks = {
					Enabled: {
						"Preparation day": preparationDay,
						"Banner up": bannerUp,
						"During the Sabbath": duringSabbath,
						"After Sabbath": afterSabbath
					}
				}
	
				console.group(`Sunset to Sunset time checks`)
				console.table(checks)
				console.groupEnd()
	
				times['Banner up'] = {
					'Date': getMessageTime(closing).toLocaleString(DateTime.DATE_FULL),
					'Time': getMessageTime(closing).toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET),
				}
	
				times['Closing guard'] = {
					'Date': closing.toLocaleString(DateTime.DATE_FULL),
					'Time': closing.toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET),
				}
	
				times['Closing sunset'] = {
					'Date': closingSunset.toLocaleString(DateTime.DATE_FULL),
					'Time': closingSunset.toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET),
				}
	
				times['Opening sunset'] = {
					'Date': openingSunset.toLocaleString(DateTime.DATE_FULL),
					'Time': openingSunset.toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET),
				}
	
				times['Opening guard'] = {
					'Date': opening.toLocaleString(DateTime.DATE_FULL),
					'Time': opening.toLocaleString(DateTime.TIME_WITH_SHORT_OFFSET),
				}

				console.group(`Sunset to Sunset times`)
				console.table(times)
				console.groupEnd()
  
			}

		})
	} else {
		console.log('Sunset to Sunset: Exiting because today is not closing day')
	}

	return options
})()